File: 2020/day02/puzzle2.ts
const input = await Deno.readTextFile("puzzle2.in");

class Password {
  private password: string;
  private numberA: number;
  private numberB: number;
  private letter: string;

  constructor(line: string) {
    const [policy, password] = line.split(": ");
    const [numbers, letter] = policy.split(" ");
    const [numberA, numberB] = numbers.split("-");

    this.password = password;
    this.numberA = parseInt(numberA);
    this.numberB = parseInt(numberB);
    this.letter = letter;
  }

  public get isPart1Valid(): boolean {
    const count = Array.from(this.password).filter((letter) => letter === this.letter).length;
    return this.numberA <= count && count <= this.numberB;
  }

  public get isPart2Valid(): boolean {
    const numberAMatches = this.password[this.numberA - 1] === this.letter;
    const numberBMatches = this.password[this.numberB - 1] === this.letter;

    return numberAMatches != numberBMatches;
  }
}

const passwords = input.split("\n").map((line) => new Password(line));

console.log(passwords.filter((password) => password.isPart1Valid).length);
console.log(passwords.filter((password) => password.isPart2Valid).length);

File: 2020/day02/puzzle2.py
file = open('puzzle2.in')

class Password:
  def __init__(self, line):
    policy, password = line.split(': ')
    numbers, letter = policy.split(' ')
    numberA, numberB = numbers.split('-')
    self.password = password
    self.numberA = int(numberA)
    self.numberB = int(numberB)
    self.letter = letter

  def is_part1_valid(self):
    count = 0

    for letter in self.password:
      if letter == self.letter:
        count += 1

    return self.numberA <= count <= self.numberB

  def is_part2_valid(self):
    numberAMatches = self.password[self.numberA - 1] == self.letter
    numberBMatches = self.password[self.numberB - 1] == self.letter

    return numberAMatches != numberBMatches

count_part1 = 0
count_part2 = 0

for line in file:
  password = Password(line)
  if password.is_part1_valid():
    count_part1 += 1
  if password.is_part2_valid():
    count_part2 += 1

print(count_part1)
print(count_part2)